# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/streams/ReadableStream.h
# Commit: fd13db3ea2b3
# Full Hash: fd13db3ea2b3bc2c96ef541680dd44f61aa2ce30
# Author: Tom Schuster <evilpies@gmail.com>
# Date: 2022-12-01 16:18:29
# Regressor Bug: 1734244
# File Overlap Count: 1
# Description:
#   Bug 1734244 - Implement async iteration of ReadableStream. r=mgaudet,peterv
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D153312
# ==============================================================================

diff -r d7c2c8133258 -r fd13db3ea2b3 dom/streams/ReadableStream.h
--- a/dom/streams/ReadableStream.h	Thu Dec 01 06:11:07 2022 +0200
+++ b/dom/streams/ReadableStream.h	Wed Nov 30 22:26:28 2022 +0000
@@ -12,6 +12,7 @@
 #include "mozilla/Attributes.h"
 #include "mozilla/ErrorResult.h"
 #include "mozilla/dom/BindingDeclarations.h"
+#include "mozilla/dom/IterableIterator.h"
 #include "mozilla/dom/QueuingStrategyBinding.h"
 #include "mozilla/dom/ReadableStreamController.h"
 #include "mozilla/dom/ReadableStreamDefaultController.h"
@@ -26,6 +27,7 @@
 class ReadableStreamDefaultReader;
 class ReadableStreamGenericReader;
 struct ReadableStreamGetReaderOptions;
+struct ReadableStreamIteratorOptions;
 struct ReadIntoRequest;
 class WritableStream;
 struct ReadableWritablePair;
@@ -143,6 +145,25 @@
                               nsTArray<RefPtr<ReadableStream>>& aResult,
                               ErrorResult& aRv);
 
+  struct IteratorData {
+    void Traverse(nsCycleCollectionTraversalCallback& cb);
+    void Unlink();
+
+    RefPtr<ReadableStreamDefaultReader> mReader;
+    bool mPreventCancel;
+  };
+
+  using Iterator = AsyncIterableIterator<ReadableStream>;
+
+  void InitAsyncIteratorData(IteratorData& aData, Iterator::IteratorType aType,
+                             const ReadableStreamIteratorOptions& aOptions,
+                             ErrorResult& aRv);
+  MOZ_CAN_RUN_SCRIPT already_AddRefed<Promise> GetNextIterationResult(
+      Iterator* aIterator, ErrorResult& aRv);
+  MOZ_CAN_RUN_SCRIPT already_AddRefed<Promise> IteratorReturn(
+      JSContext* aCx, Iterator* aIterator, JS::Handle<JS::Value> aValue,
+      ErrorResult& aRv);
+
   // Internal Slots:
  private:
   RefPtr<ReadableStreamController> mController;